File: codeeditor/src/main/java/com/mcal/neweditor/editor2/smali/SmaliMethodAdapter.java
package com.mcal.neweditor.editor2.smali;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.mcal.neweditor.R;

import java.util.List;

public class SmaliMethodAdapter extends BaseAdapter {
    private final Context ctx;
    private final List<SmaliMethodInfo> methods;

    public SmaliMethodAdapter(Context ctx, List<SmaliMethodInfo> methods) {
        this.ctx = ctx;
        this.methods = methods;
    }

    @Override
    public int getCount() {
        return methods.size();
    }

    @Override
    public Object getItem(int position) {
        return methods.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder viewHolder;
        if (convertView == null) {
            convertView = LayoutInflater.from(ctx).inflate(R.layout.popup_item_small, null);
            viewHolder = new ViewHolder();
            viewHolder.tv = convertView.findViewById(R.id.groupItem);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        // Set the method description
        viewHolder.tv.setText(methods.get(position).methodDesc);

        return convertView;
    }

    static class ViewHolder {
        public TextView tv;
    }
}

File: codeeditor/src/main/java/com/mcal/editor/smali/SmaliMethodAdapter.kt
package com.mcal.editor.smali

import android.graphics.Color
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.mcal.neweditor.R

class SmaliMethodAdapter(val listener: OnClick, private val methods: List<SmaliMethodInfo>) : RecyclerView.Adapter<SmaliMethodAdapter.SmaliViewHolder>() {
    override fun getItemCount(): Int {
        return methods.size
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SmaliViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.popup_item_small, parent, false)
        return SmaliViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: SmaliViewHolder, position: Int) {
        val text = methods[position].methodDesc
        val spanText = SpannableString(text)
        if (text.contains("(")) {
            holder.type.text = "M"
            holder.type.setBackgroundColor(Color.parseColor("#FFAB91"))
            val start = text.indexOf("(")
            val end = text.indexOf(")") + 1
            if (start >= 0 && end >= 0) {
                spanText.setSpan(ForegroundColorSpan(Color.GRAY), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            holder.tv.text = spanText
        } else {
            holder.type.text = "F"
            holder.type.setBackgroundColor(Color.parseColor("#B39DDB"))
            val start = text.indexOf(":")
            val end = text.indexOf(";")
            if (start >= 0 && end >= 0) {
                spanText.setSpan(ForegroundColorSpan(Color.GRAY), start, end, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            holder.tv.text = spanText
        }
        holder.tv.setOnClickListener {
            listener.onClick(position)
        }
    }

    class SmaliViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var tv: TextView = itemView.findViewById(R.id.groupItem)
        var type: TextView = itemView.findViewById(R.id.type)
    }
}